### PR TITLE
[npm] find modules in npm cache

### PR DESCRIPTION
**Pull request by TomasTomecek**
npm installs modules:
1. globally — so everyone can use those
2. per project

If you don't know the path to the project, the only way to see the list
of those modules is by traversing the whole filesystem, which is pretty
unreal.

This patch browses npm cache of every user on the system and saves
module names and all available versions into sosreport.

### CODE REVIEW

**Comment by bmr-cymru** [comment]
This seems like it is papering over a deficiency in `npm`; if it doesn't provide a mechanism to obtain this "global list" then it isn't really sos's place to address that.

I think it would be better to take this to the `npm` developers and request a formal mechanism for obtaining this information.

We do not have a precedent in `sos` for parsing `/etc/passwd` and trawling through user's home directories (we collect a few fixed items from `/root`, but that is it); the only other "home directory" collection is for specific, named (system) accounts, e.g. `nocpulse` and `stack`.

Since the commit acknowledges that the format of `~/.npm` is subject to change, we don't have any good way to validate what we find here - I know it is unlikely, but this patch could cause us to collect arbitrary data from user's home directories that is completely unrelated to the `npm` package.

**Comment by TomasTomecek** [comment]
It's very true what you wrote. I already created such [feature request](https://github.com/npm/npm/issues/13510), unfortunately received zero response.

I completely understand your arguments, will pass your feedback to our team. Thanks for such a quick response.

**Comment by bmr-cymru** [comment]
OK, thanks for the link to the `npm` request - let's give it some time, to see if that might still go somewhere useful, and if not we can look at how best we can work around that in `sos`?

**Comment by TomasTomecek** [comment]
> if not we can look at how best we can work around that in `sos`?

Sounds okay. To be honest, this is the most easiest & least intrusive workaround (scanning whole filesystem was pretty no-go).

**Comment by bmr-cymru** [comment]
Honestly, if `npm` cannot help us here, then it is more thank likely it will be either:

* Add an option, `npm.npm_users`, that accepts a list of `npm` using accounts whose home directories should be instpected,

OR:

* If sos is being run as a normal user, inspect the `~/.npm` of that account.

I realise these do not get the same comprehensive data collection as your proposal, however, given these are user home directories and we have literally no clue what is in them, I think this is as far as I would be willing to go in `sos` without explicit support from `npm`.

File: sos/plugins/npm.py [hunk 1]
```diff
@@ -44,13 +46,49 @@ def _get_npm_output(self, cmd, filename, working_directory=None):
             runat=working_directory
         )
 
+    def _find_modules_in_npm_cache(self):
+        """
+        since format of `~/.npm` is changing with newer npm releases, let's do
+        the easiest thing and don't predict nor expect anything, the structure
+        is suppose to be:
```
Comment (jpopelka) [inline sos/plugins/npm.py]: I've tried npm v1, v2, v3 and latest upstream v4 and the structure as outlined here hasn't changed.
The only thing that changed recently is where the `.cache.json`s are stored - previously they were in `.npm/name/` while now they are in separate `.npm/registry.npmjs.org/name/`. Is that what you meant in the comment Tomas ?

**Comment by jpopelka** [comment]
> Honestly, if npm cannot help us here, then it is more than likely it will be either:
>     Add an option, npm.npm_users, that accepts a list of npm using accounts whose home directories should be inspected,

I don't know much about how sosreport is used in real life, but I doubt anybody would bother using such option.

> OR:
>     If sos is being run as a normal user, inspect the ~/.npm of that account.

I've created new https://github.com/sosreport/sos/pull/899 which implements this.

**Comment by bmr-cymru** [comment]
I have merged this as-is for now - with the patch from @jpopelka to use `npm` for the cache lookup I'm happy with the home directory trawling.

**Pull request closed without merging**